includes/data-events/class-woo-user-registration.php

```diff
@@ -0,0 +1,99 @@
+<?php
+/**
+ * Newspack Data Events Woo Registration hooks.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack\Data_Events;
+
+/**
+ * Class that triggers a registration event with Newspack metadata for users registered during the Woocommerce checkout process.
+ */
+final class Woo_User_Registration {
+
+	/**
+	 * Whether the current request is processing a checkout.
+	 *
+	 * @var boolean
+	 */
+	private static $processing_checkout = false;
+
+	/**
+	 * Metadata to send with the registration event.
+	 *
+	 * @var array
+	 */
+	private static $metadata = [];
+
+	/**
+	 * Initialize the class.
+	 *
+	 * @return void
+	 */
+	public static function init() {
+		// is processing checkout?
+		add_action( 'woocommerce_checkout_process', [ __CLASS__, 'checkout_process' ] );
+
+		// created a user?
+		add_action( 'woocommerce_created_customer', [ __CLASS__, 'created_customer' ], 1 );
+	}
+
+	/**
+	 * During the checkout process, set the processing_checkout flag and store metadata.
+	 *
+	 * @return void
+	 */
+	public static function checkout_process() {
+
+		// If a user is created later on the request by woocommerce_created_customer(), it will at least have this data.
+		self::$metadata['registration_method'] = 'woocommerce';
+
+		/**
+		 * On Newspack\Donations::process_donation_form(), we add these values to the cart.
+		 *
+		 * Later, we add them to the order (Newspack\Donations::checkout_create_order_line_item()) and use it to send the metadata to Newspack on donation events.
+		 *
+		 * Here, we are going to read the same information from the cart and use it to send the metadata to Newspack on registration events.
+		 */
+		foreach ( \WC()->cart->get_cart() as $cart_item_key => $values ) {
+			if ( ! empty( $values['newspack_popup_id'] ) ) {
+				self::$metadata['newspack_popup_id'] = $values['newspack_popup_id'];
+			}
+			if ( ! empty( $values['referer'] ) ) {
+				self::$metadata['referer'] = $values['referer'];
+			}
+		}
```

Comment:
This data will only exist for Newspack donations.. maybe we could then set registration_method to "newspack_donation" instead of a generic "woocommerce" ?

Reply:
Would the registration_method still be `woocommerce` if it's a premium subscription purchase (not donation)?

Reply:
yes. wdyt?

Reply:
I could add `\WC()->cart->get_cart()` as a metadata to the event.

But this could also be fetched by the integration directly from Woocommerce if needed. No need to attach it to the event...

Reply:
I think having the `registration_method` consolidated as `woocommerce` whenever WC is registering makes more sense to me, but I don't mind if you'd rather have it as `newspack_donation` for this case and `woocommerce` for every other.

Reply:
I agreed to keep it woocommerce in every case. I was just pointing that we don't need to add additional metadata. If we want to look for the products in the cart in a particular integration we can look directly into the cart itself :)